library/common/config/config.cc

```diff
@@ -280,6 +271,46 @@ const char* config_template = R"(
             typed_config:
               "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
 
+!ignore tls_socket_defs:
+- &validation_context {{custom_cert_validation_context}}
```

Comment:
The main config (`config_header` + `config_template`) concatenated is 100% valid YAML and a valid default configuration for Envoy Mobile. Adding this breaks that convention, since {{'s syntactical meaning in YAML. This is why the few places where interpolation via token still occurs in the main config, the token uses a style that's interpreted by YAML as a comment (`#{}`) and the config is valid when it's empty.

Reply:
Hmm, if I change it to use `#{custom_cert_validation_context}` but without replacement with `platform_cert_validation_context_template` or `default_cert_validation_context_template` defined above, the YAML will have an empty `validation_context:`, which is a valid YAML, but not a good TLS config for Envoy.

I see that `{{`s are used in template inserts, i.e. http filters, which can be optional, but I don't think validation context should be optional, and the engine has to config it one way or the other. So the question here is whether it still meaningful to make the config_template a valid default configuration, given the engine will do the replacement all the time?

Reply:
Can you explain a bit more about this? Are you suggesting to have 2 `&validation_context` definitions and somehow the added one can take effect? And why they have to be in the config_header? AFAIK, config_header string is not manipulated by the engine builder at all.

Reply:
Sure, let me try to explain my understanding of our YAML setup.

So when you have more than one 'definition' (YAML anchor) such as `&validation_context` the last processed one takes effect when we evaluate `*validation_context` (YAML alias). An example from some YAML playground that I use to play with it in the past ([link](https://eemeli.org/yaml-playground/)):

<img width="579" alt="Screen Shot 2022-09-27 at 9 13 41 AM" src="https://user-images.githubusercontent.com/4410346/192536240-63a22f3c-6555-4e9b-9816-7fa47bf78365.png">

Now about the `config_header`. Its main purpose is to act as a place for all of the default values for various YAML variables (variable is not the right YAML term, the right term is YAML *anchor*) so what we could use it for is to provide a default value for `validation_context`.

It is fine that that engine builder does *not* modify the `config_header`. The engine builder does prepend content to the `config_template` part of the config so the final Envoy config looks something as follows:

```
1. [CONTENT OF CONFIG_HEADER]
2. [CONTENT ADDED BY THE ENGINE BUILDER] (it's prepended to the config_template in [here](https://github.com/envoyproxy/envoy-mobile/blob/main/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java#L306))
3. [CONTENT OF CONFIG_TEMPLATE] 
```

So looking at ^ you can provide a default value in 1):
```
- &validation_context R"(
  trusted_ca:
    inline_string: *tls_root_certs)";
```
 
override it in 2) as needed (conditionally) - by either adding or not adding the following string:

```
- &validation_context R"(
  custom_validator_config:
    name: "envoy_mobile.cert_validator.platform_bridge_cert_validator")"
```

and this is what will be visible to YAML in 3).

Reply:
>when we don't config platform cert validator, even though it's duplicated with config_header. This is because &trust_chain_verification gets overridden.

Ah, I see - so it's a little bit more tricky here. Nice catch 👍

Reply:
After a little bit more of thinking. Should we just move `validation_context_config_trust_chain ` anchor from `config_header` to the beginning of the `config_template` instead and keep the `validation_context` where it is now? That should work I think (and the resulting YAML would still be a valid config)?

Reply:
I deprecated `validation_context` and used `validation_context_config_trust_chain` for all protocols. The templates are consolidated to 2 right now, default validation v.s. platform-based validation. PTAL.